### PR TITLE
Fix log decoding in contract tests

### PR DESCRIPTION
**Pull request by abandeali1**
## Description

- Makes log decoding in contract tests independent from 0x.js (this was causing tests to fail).
- Moves `utils` directory into `src` for consistency. 

## Types of changes

* [x] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

* [ ] Change requires a change to the documentation.
* [ ] Added tests to cover my changes.
* [ ] Added new entries to the relevant CHANGELOGs.
* [ ] Updated the new versions of the changed packages in the relevant CHANGELOGs.
* [ ] Labeled this PR with the 'WIP' label if it is a work in progress.
* [ ] Labeled this PR with the labels corresponding to the changed package.

### CODE REVIEW

File: packages/contracts/src/artifacts/MixinExchangeCore.json [hunk 1]
```diff
@@ -1,5 +1,4 @@
 {
-    "contract_name": "MixinExchangeCore",
```
Comment (fabioberger) [inline packages/contracts/src/artifacts/MixinExchangeCore.json]: Why was this removed?

File: packages/contracts/src/utils/exchange_wrapper.ts [hunk 2]
```diff
@@ -35,7 +35,11 @@ export class ExchangeWrapper {
         );
         const tx = await this._zeroEx.awaitTransactionMinedAsync(txHash);
         tx.logs = _.filter(tx.logs, log => log.address === this._exchange.address);
-        _.each(tx.logs, log => wrapLogBigNumbers(log));
+        tx.logs = _.map(tx.logs, log => {
+            const decodedLog = this._logDecoder.tryToDecodeLogOrNoop(log);
```
Comment (fabioberger) [inline packages/contracts/src/utils/exchange_wrapper.ts]: So this is not correct handling of the return value of the method, **unless** you are absolutely certain that given your log, it will always return the decoded log. Are you?

File: packages/contracts/src/utils/log_decoder.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import { AbiDecoder } from '@0xproject/utils';
+import * as _ from 'lodash';
+import * as Web3 from 'web3';
+
+import { artifacts } from './artifacts';
+import { Artifact } from './types';
+
+export class LogDecoder {
+    private _abiDecoder: AbiDecoder;
+    constructor(networkId: number) {
```
Comment (fabioberger) [inline packages/contracts/src/utils/log_decoder.ts]: This should be `networkIdIfExists?: number`, if the value can every be `undefined`.

File: packages/contracts/src/utils/log_decoder.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        if (_.isUndefined(networkId)) {
+            throw new Error('networkId not specified');
+        }
+        const abiArrays: Web3.AbiDefinition[][] = [];
+        _.forEach(artifacts, (artifact: Artifact) => {
+            const network = artifact.networks[networkId];
```
Comment (fabioberger) [inline packages/contracts/src/utils/log_decoder.ts]: rename to `networkIfExists`

File: packages/contracts/test/exchange/core.ts [hunk 5]
```diff
@@ -430,7 +425,8 @@ describe('Exchange', () => {
             });
             expect(res.logs).to.have.length(1);
 
-            const logArgs = (res.logs[0] as LogWithDecodedArgs<LogFillContractEventArgs>).args;
+            const log = res.logs[0] as LogWithDecodedArgs<LogFillContractEventArgs>;
+            const logArgs = log.args;
```
Comment (fabioberger) [inline packages/contracts/test/exchange/core.ts]: 💯

File: packages/contracts/test/exchange/core.ts [hunk 6]
```diff
@@ -526,153 +523,149 @@ describe('Exchange', () => {
...
-        it('should not change balances if maker balances are too low to fill order and \
-                shouldThrowOnInsufficientBalanceOrAllowance = false', async () => {
-            signedOrder = await orderFactory.newSignedOrderAsync({
-                makerTokenAmount: ZeroEx.toBaseUnitAmount(new BigNumber(100000), 18),
-            });
+        // it('should not change balances if maker balances are too low to fill order and \
+        //         shouldThrowOnInsufficientBalanceOrAllowance = false', async () => {
+        //     signedOrder = await orderFactory.newSignedOrderAsync({
+        //         makerTokenAmount: ZeroEx.toBaseUnitAmount(new BigNumber(100000), 18),
+        //     });
```
Comment (fabioberger) [inline packages/contracts/test/exchange/core.ts]: Is this meant to remain commented out? Lot's of that in this file.

File: packages/contracts/src/utils/exchange_wrapper.ts [hunk 7]
```diff
@@ -3,13 +3,16 @@ import { BigNumber } from '@0xproject/utils';
...
+import { ExchangeContract } from '../contract_wrappers/generated/exchange';
 
+import { constants } from './constants';
 import { formatters } from './formatters';
+import { LogDecoder } from './log_decoder';
 import { signedOrderUtils } from './signed_order_utils';
 
 export class ExchangeWrapper {
     private _exchange: ExchangeContract;
+    private _logDecoder: LogDecoder = new LogDecoder(constants.TESTRPC_NETWORK_ID);
```
Comment (fabioberger) [inline packages/contracts/src/utils/exchange_wrapper.ts]: Do you expect to instantiate `logDecoder` in multiple places?

File: packages/contracts/src/artifacts/MixinExchangeCore.json (hunk 1 above)
Comment (abandeali1) [reply packages/contracts/src/artifacts/MixinExchangeCore.json]: Hmm very strange, no idea.

File: packages/contracts/src/utils/log_decoder.ts (hunk 3 above)
Comment (dekz) [reply packages/contracts/src/utils/log_decoder.ts]: I think since this is an immediate validation and throw that `networkIdIfExists?` isn't necessary.  Perhaps if undefined here was valid option and could be handled, not thrown.

File: packages/contracts/src/utils/exchange_wrapper.ts (hunk 7 above)
Comment (abandeali1) [reply packages/contracts/src/utils/exchange_wrapper.ts]: Probably not, why?

File: packages/contracts/test/exchange/core.ts (hunk 6 above)
Comment (abandeali1) [reply packages/contracts/test/exchange/core.ts]: Yeah, tests that are broken are commented out for now (they need to be ported over to `fillOrderNoThrow` still).

File: packages/contracts/src/artifacts/MixinExchangeCore.json (hunk 1 above)
Comment (fabioberger) [reply packages/contracts/src/artifacts/MixinExchangeCore.json]: Let's undo.

File: packages/contracts/src/utils/log_decoder.ts (hunk 3 above)
Comment (fabioberger) [reply packages/contracts/src/utils/log_decoder.ts]: There is no guarentee this validation will stay in place. I want the semantic meaning of this arg to be: "hey, watch out, this might be passed in as undefined".

File: packages/contracts/src/utils/exchange_wrapper.ts (hunk 7 above)
Comment (fabioberger) [reply packages/contracts/src/utils/exchange_wrapper.ts]: Because if so, there would be no need for the `LogDecoder` class. But perhaps it's cleaner this way.

File: packages/contracts/src/utils/log_decoder.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        _.forEach(artifacts, (artifact: Artifact) => {
+            const networkIfExists = artifact.networks[networkIdIfExists];
+            if (_.isUndefined(networkIfExists)) {
+                throw new Error(`Artifact does not exist on network ${networkIdIfExists}`);
+            }
+            abiArrays.push(networkIfExists.abi);
+        });
+        this._abiDecoder = new AbiDecoder(abiArrays);
+    }
+    public tryToDecodeLogOrNoop<ArgsType>(log: Web3.LogEntry): LogWithDecodedArgs<ArgsType> | RawLog {
```
Comment (fabioberger) [inline packages/contracts/src/utils/log_decoder.ts]: Idea: Let's make the public interface to this class mirror what we eventually want the interface for AbiDecoder to look like. Let's change this method to be:

```
public decodeLogOrThrow<ArgsType>(log: Web3.LogEntry): LogWithDecodedArgs<ArgsType> {
  const logWithDecodedArgsOrLog = this._abiDecoder.tryToDecodeLogOrNoop(log);
  if (_.isEqual(logWithDecodedArgsOrLog, log)) {
    throw new Error(`Unable to decode log: ${JSON.stringify(log)}`);
  }
  return logWithDecodedArgsOrLog;
}
```
